src/lib/ares_process.c

```diff
@@ -71,35 +71,66 @@ static void          end_query(ares_channel_t *channel, struct query *query,
                                ares_status_t status,
                                const ares_dns_record_t *dnsrec);
 
-static void          server_increment_failures(struct server_state *server)
+static void server_increment_failures(struct server_state *server)
 {
-  ares__slist_node_t   *node;
-  const ares_channel_t *channel = server->channel;
+  ares__slist_node_t *node;
+  ares_channel_t     *channel = server->channel;
+  char                server_ip[INET6_ADDRSTRLEN];
 
   node = ares__slist_node_find(channel->servers, server);
   if (node == NULL) {
     return;
   }
+
+  server->consec_successes = 0;
   server->consec_failures++;
   ares__slist_node_reinsert(node);
+
+  /* If the server has just become unhealthy, then invoke the server state
+   * callback.
+   */
+  if ((server->consec_failures >= channel->server_failure_threshold) &&
+      (server->is_healthy == ARES_TRUE)) {
+    server->is_healthy = ARES_FALSE;
+    if (channel->server_state_cb) {
+      ares_inet_ntop(server->addr.family, &server->addr.addr, server_ip,
+                     sizeof(server_ip));
+      channel->server_state_cb(server_ip, server->is_healthy,
+                               channel->server_state_cb_data);
+    }
+  }
```

Comment:
so `channel->server_failure_threshold` is only used to control the `is_healthy` flag, the server list still gets sorted regardless?  Seems like it may take quite a while to trigger the callback for unhealthy since you're relying on a good number of queries with `priority_server_chance` to trigger this.  Maybe that's ok, but would definitely need to be documented.

Reply:
To give a bit of context, this change was originally written prior to the sorted server list, when:
- Servers were stored in an array on the channel.
- The server to send the next request to was stored in the `query` structure (`query->server`).
- If a request needed to be retried, we moved onto the next server with `query->server = (query->server + 1) % channel->nservers`.

This change still tracked healthy/unhealthy servers in the same way (based on consecutive failures/successes). But unhealthy servers were skipped at the point when moving onto the next server. E.g. when we moved to the next server, if it was unhealthy then we'd skip it with 90% probability.

This doesn't quite work now, because we always use the first server in the sorted list on retries. For example if we sorted by `(is_healthy, idx)`, then retries would always go to the first server until it hits 3 failures and gets marked as unhealthy. But I think we want retries to cycle through the servers in the list.

Reply:
Perhaps the following would work:

- Introduce a new field on servers called `requests_for_current_query` that tracks how many times the server has been used on the current query.
- This field could be incremented in `server_increment_failures()` and `server_increment_successes()`. And it would get reset to 0 on all servers before the first call to `ares__send_query()`
- Servers would then be sorted by `(is_healthy, requests_for_current_query, idx)`.

That way:
- Unhealthy servers would be deprioritised. But only after they have hit the threshold for the number of consecutive failures.
- We still move to new servers when retrying requests within the same query.
- The highest priority healthy server will be used on the first attempt in a query. (We could still have the same logic to retry higher priority unhealthy servers with a small probability.)

Reply:
sounds a little complex for what is really needed, and we can't sort the server list by requests_for_current_query since there can be multiple outstanding queries to a single server server at one time.

I do think it makes sense to have logic to retry unhealthy servers with a small probability, however it probably makes sense to have a minimum time for such a try as otherwise for applications which perform a large number of DNS queries (thinking here, something doing a lot of reverse lookups in a webserver log to dereference an ip address to a name), that would end up stalling a lot if that server was still down.  So we'd want to track both down time and if its actively being attempted to see if its still live by another query incase it 'hits' again.

And I really do think we need to remove the is_healthy concept along with number of tries for up or down, as we can leave the actual alert logic to the application.

Reply:
Hi Brad, thank you for the feedback and apologies for the delay in getting the markups. I have pushed some further changes based on the feedback. A summary:

Server state callback:
- I removed the concept of a healthy/unhealthy server (application code can handle decisions like this).
- The callback now always fires on success or failure (with the `ares_bool_t` argument indicating this).
- The server string argument now includes extra details like the port and interface.

Server failover retry behaviour:
- This is now controlled by 3 fields:
  - `retry_chance` - The probability that we retry a failed server.
  - `retry_delay` - The minimum delay between retries (channel-wide delay, rather than per-server).
  - `serious_fail_limit` - The number of consecutive failures beyond which a server is considered to have a serious failure.
    - If all servers exceed this, then servers are always selected at random.
    - Moreover, connections to such servers are always closed out.
- When sending queries, if the `retry_delay` has passed and the `retry_chance` is hit, then we select the first failed server in the server list.
- We no longer consider "higher priority" failed servers. This means that all failed servers will eventually be retried.

I still need to test this and there are some optimisations I can still make, but the high level behaviour should be OK for re-review. Please let me know if you have further thoughts. Thanks!